### PR TITLE
8280139: Report more detailed statistics about task stealing in task queue stats 

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews that improves task queue statistics?

This is a step to see problems with task queue in general and measure future task stealing improvements in GC; particularly contention on steal information has shown to be correlated to performance, so information about basic occurrences during stealing (how many empty, how many contended, how many contended in a row, max contended accesses in a row) in the task queue is very interesting for further performance work.

Old output of `gc+tasks+stats=trace` after compiling in these statistics:
```
thr      qpush       qpop     qpop-s   qattempt     qsteal      opush       omax
--- ---------- ---------- ---------- ---------- ---------- ---------- ----------
  0    3777115    3776816       3872      17434       9630          0          0
```
And new:
```
thr        push         pop    pop-slow  st-attempt    st-empty     st-ctdd  st-success st-ctdd-max st-biasdrop  ovflw-push   ovflw-max
--- ----------- ----------- ----------- ----------- ----------- ----------- ----------- ----------- ----------- ----------- -----------
  0     4616264     4616218        3608       12945           1        1650       11294           5        9405           0           0
[...]
```
The first four columns stayed the same, then the `qsteal` has been split up into the various steal results (empty/contended/success). `st_ctdd_max` gives the maximum amount of contended steal attempts in a row. `st_biasdrop` shows how many times the steal bias has been dropped (reset).
`ovflw-push` and `ovfl-max` are the same as before, just renamed a bit to use the available space.

There has been no intention to change anything about the algorithm, so there is no need for perf testing (I can see). Later changes will do that.

Thanks,
  Thomas

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280139](https://bugs.openjdk.java.net/browse/JDK-8280139): Report more detailed statistics about task stealing in task queue stats

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7143/head:pull/7143` \
`$ git checkout pull/7143`

Update a local copy of the PR: \
`$ git checkout pull/7143` \
`$ git pull https://git.openjdk.java.net/jdk pull/7143/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7143`

View PR using the GUI difftool: \
`$ git pr show -t 7143`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7143.diff">https://git.openjdk.java.net/jdk/pull/7143.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shared/taskqueue.hpp [hunk 1]
```diff
@@ -357,7 +362,13 @@ class GenericTaskQueue: public TaskQueueSuper<N, F> {
 
   // Like pop_local(), but uses the "global" end of the queue (the least
   // recently pushed).
-  bool pop_global(E& t);
+  // The result value order of this correspond to the order in the corresponding StatId.
+  enum PopResult {
```
[the remark this answers is not in the document]
Comment (kimbarrett) [reply src/hotspot/share/gc/shared/taskqueue.hpp:366]: Also, consider making `PopResult` an enum class.

File: src/hotspot/share/gc/shared/taskqueue.inline.hpp [hunk 2]
```diff
@@ -291,17 +293,19 @@ GenericTaskQueueSet<T, F>::steal_best_of_2(uint queue_num, E& t) {
     uint sz2 = _queues[k2]->size();
 
     uint sel_k = 0;
-    bool suc = false;
+    PopResult suc = T::Empty;
 
     if (sz2 > sz1) {
       sel_k = k2;
       suc = _queues[k2]->pop_global(t);
+      TASKQUEUE_STATS_ONLY(queue(queue_num)->stats.record_steal_attempt((uint)suc);)
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/taskqueue.inline.hpp:301]: Consider adding `TaskQueueSuper::record_steal_attempt(PopResult)` to package up this idiom.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/shared/taskqueue.inline.hpp [hunk 3]
```diff
@@ -205,7 +205,7 @@ bool OverflowTaskQueue<E, F, N>::pop_overflow(E& t)
 // reads elems[oldAge.top].  The owner's bottom == the thief's oldAge.top.
 // (4) Thief will discard the read value, because its cmpxchg of age will fail.
 template<class E, MEMFLAGS F, unsigned int N>
-bool GenericTaskQueue<E, F, N>::pop_global(E& t) {
+typename GenericTaskQueue<E, F, N>::PopResult GenericTaskQueue<E, F, N>::pop_global(E& t) {
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/taskqueue.inline.hpp:208]: These are places that are just made for trailing return types, putting the return type in the scope of the function.  That's still in the "undecided features" list in the style guide though.

```
template<class E, MEMFLAGS F, unsigned int N>
auto GenericTaskQueue<E, F, N>::pop_global(E& t) -> PopResult { ... }
```

**Review by kimbarrett — APPROVED** [review]

**Comment by tschatzl** [comment]
In need for a second reviewer... :)

**Review by walulyai — APPROVED** [review]
Lgtm!

**Comment by tschatzl** [comment]
Thanks @kimbarrett @walulyai for your reviews

/integrate

**Pull request closed without merging**